### ISSUE TITLE
question about grid search options

### ISSUE THREAD

**Issue by xiaolongwu0713** [skorch-dev/skorch#762]
Hi, I have some questions.
1, I notice in the doc that gridSearchCV will handle the training/validating split, however I used to use train_split=predefined_split(val_ds) to do my own train/valid split. So in the grid search situation, is it means I have to stop splitting myself? If so, I assume I should put both my original train and validate set  as new training set and feed to the grid search?
2, can I do the grid search on different optimizor, for example: 
```
params = {
    'lr': [0.05, 0.1],
    'optimizer': [torch.optim.Adam, torch.optim.Adagrad,torch.optim.SGD]
}
```
3, Is there any way I can extract all my tunable parameter form the network so that I can have a big picture of what grid search option I have?

Many thanks.

**Comment by BenjaminBossan** [comment]
> it means I have to stop splitting myself?

Exactly, grid search works in this way. On the skorch side, set `train_split=None` and probably you should also set `verbose=0`.

> If so, I assume I should put both my original train and validate set as new training set and feed to the grid search?

If you still have a completely separate test set, you can throw your training and validation set together, otherwise it might be better to just use the training set. Depending on the kind of task you're solving, you have to decide which CV splitting strategy to use with grid search.

>  can I do the grid search on different optimizor, for example:

Yes, absolutely.

> 3, Is there any way I can extract all my tunable parameter form the network so that I can have a big picture of what grid search option I have?

Yes: `net.get_params()` returns all parameters you can possibly set. However, this does not include parameters on the PyTorch stuff like the module or optimizer. So e.g. if you use `torch.optim.Adam`, it will not list that `optimizer__betas` is a possible argument. Same for the module, data loader etc.

**Comment by xiaolongwu0713** [comment]
Thanks for your answer.
Do I need to initialize my network if I do grid search? for example, I define my net like below:
```
net = MyRegressor(
    module=TSception2,
    module__wind_size=100,
    module__step=10,
    module__sampling_rate=1000,
    module__chnNum=2,
    module__num_T=2,
    module__num_S=2,
    module__batch_size=1,
    module__dropout=0.5,
    #train_split=predefined_split(valid_set),
    iterator_train__shuffle=True,
    #train_split=predefined_split(val_ds), #disable this when do grid search
    max_epochs=200,
    lr=learning_rate,
    batch_size=1,
    optimizer=torch.optim.Adam,
    criterion = nn.MSELoss,
    callbacks=[('plotPrediction', plotPrediction(result_dir)),cp,train_end_cp],
    device = device
)
```

Then I define my parameter grid:
```
params = {
    'lr': [0.001, 0.002, 0.003],
    'max_epochs': [100,200, 300],
    'module__dropout': [0.1,0.2,0.3,0.4,0.5],
}
```

As you can see, I define the dropout parameter in two places, net definition and the params. But I think grid search would choose the parameter in its own fashion, so I guess there is no point to initialize them in the net definition?

**Comment by BenjaminBossan** [comment]
> As you can see, I define the dropout parameter in two places, net definition and the params. But I think grid search would choose the parameter in its own fashion, so I guess there is no point to initialize them in the net definition?

Exactly, the grid search will override the parameters that you used for initialization.

> Should this scoring function be the same as the criterion defined in the net definition? There is no reason why they should be different, right?

There can be legitimate reasons to choose a scoring function that is different from your criterion. E.g.:

- your scoring function is non-differentiable 
- the derivative of your scoring function is not well behaved or slow to calculate
- you want to pass multiple scoring functions
- you want to optimize a score that is easy to explain to your stakeholders but is not a good loss function (e.g. accuracy)

You should familiarize yourself well with your problem so that you can determine if any of these conditions apply to it.

**Comment by xiaolongwu0713** [comment]
That's a very clear answer. Thansks.
